presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTokenSplitManager.java

```diff
@@ -13,131 +13,192 @@
     {
-        this.cassandraThriftClient = new CassandraThriftClient(requireNonNull(connectionFactory, "connectionFactory is null"));
-        this.executor = requireNonNull(executor, "executor is null");
-        this.splitSize = config.getSplitSize();
-        try {
-            this.partitioner = FBUtilities.newPartitioner(config.getPartitioner());
-        }
-        catch (ConfigurationException e) {
-            throw new RuntimeException(e);
-        }
+        this(session, config.getSplitSize());
+    }
+
+    public CassandraTokenSplitManager(CassandraSession session, int splitSize)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.splitSize = splitSize;
     }
 
-    public List<TokenSplit> getSplits(String keyspace, String columnFamily)
-            throws IOException
+    public List<TokenSplit> getSplits(String keyspace, String table)
     {
-        List<TokenRange> masterRangeNodes = cassandraThriftClient.getRangeMap(keyspace);
+        Set<TokenRange> tokenRanges = getTokenRanges();
+
+        if (tokenRanges.isEmpty()) {
+            throw new PrestoException(GENERIC_EXTERNAL, "Cluster metadata is not available. " +
+                    "Please make sure that Cassandra cluster is up and running, " +
+                    "and the contact points specified are correct");
+        }
+
+        if (tokenRanges.stream().anyMatch(TokenRange::isWrappedAround)) {
+            tokenRanges = unwrap(tokenRanges);
+        }
+
+        Optional<TokenRing> tokenRing = createForPartitioner(getPartitioner());
+        long totalPartitionsCount = getTotalPartitionsCount(keyspace, table);
 
-        // canonical ranges, split into pieces, fetching the splits in parallel
         List<TokenSplit> splits = new ArrayList<>();
-        List<Future<List<TokenSplit>>> splitFutures = new ArrayList<>();
-        for (TokenRange range : masterRangeNodes) {
-            // for each range, pick a live owner and ask it to compute bite-sized splits
-            splitFutures.add(executor.submit(new SplitCallable<>(range, keyspace, columnFamily, splitSize, cassandraThriftClient, partitioner)));
+        for (TokenRange tokenRange : tokenRanges) {
+            if (tokenRange.isEmpty()) {
+                continue;
+            }
+            checkState(!tokenRange.isWrappedAround(), "all token ranges must be unwrapped at this step");
+
+            List<String> endpoints = getEndpoints(keyspace, tokenRange);
+            checkState(!endpoints.isEmpty(), "endpoints is empty for token range: %s", tokenRange);
+
+            if (!tokenRing.isPresent()) {
+                checkState(!tokenRange.isWrappedAround(), "all token ranges must be unwrapped at this step");
+                splits.add(createSplit(tokenRange, endpoints));
+                continue;
+            }
+
+            double tokenRangeRingFraction = tokenRing.get().getRingFraction(tokenRange.getStart().toString(), tokenRange.getEnd().toString());
```

Comment:
I really don't think it is worth to create 2 lines method that accepts 3 barely related parameters. I don't even know how to call it to make it readable. If you don't mind i would like to leave it as is.